### DISCUSSION TITLE
New Attribute as Sensor helper

### DISCUSSION THREAD

**Discussion by gjohansson-ST** [home-assistant/architecture#821] [General] [answered]
## Context

Suggestions from the community to get a simple helper to create a new sensor from the attribute of another sensor.
https://community.home-assistant.io/t/wth-are-attributes-such-a-pain-to-work-with/467837
https://community.home-assistant.io/t/new-helper-entity-from-attribute/461078/5

## Proposal

Create a new helper `attribute_as_sensor` that consumes the attribute of one entity to create a state for a new sensor.

Many times this is the first step of using templates for many and this would make it even simpler for the community to get such sensors going to use them both for frontend, automations etc.

PR (closed at the moment): https://github.com/home-assistant/core/pull/79835

**Comment by frenck** [comment]
Please provide some concrete examples and use cases you are trying to solve.

> Many times this is the first step of using templates for many and this would make it even simpler for the community to get such sensors going to use them both for frontend, automations etc.

Automations already support attributes and is IMHO thus not a valid reasoning.

**Comment by gjohansson-ST** [comment]
I think first and foremost this new helper would be used to "replace" the easier of templates where you have an attribute you want to monitor/see and you can also attach other properties to it such as device_class etc. to really have it as it's own entity.

For automations it's true that attribute is easily available but many as you know also use NR instead of HA's capabilities and there attributes are not as pleasantly available so you either have to script it or create template sensors to use them instead.

So I personally use it for sun elevation, next sunset, sunrise, is it daylight according to my motion sensor and quite a bunch more including some custom components (which tend to put lots in attributes) which I both put in frontend but also use in automations.

We should/could also extend this helper further in next step to allow for nested attributes such as in weather forecast as you might want to be able to show the rain for the next three days or something as such.
And also consider probably to extend it into options of binary_sensor so one can choose between creating a sensor or a binary sensor depending on what you want to do with your new entity.

Lots will still be attributes even if it's going to be even less than today eventually and this would be a simple UI friendly way to use them in a proper way.

**Reply by frenck** [reply]
> I think first and foremost this new helper would be used to "replace" the easier of templates where you have an attribute you want to monitor/see and you can also attach other properties to it such as device_class etc. to really have it as it's own entity.

This example is odd, as the device class is a static property. Extracting those in a sensor with a helper should be disallowed and won't provide any value (besides raising memory and adding more unneeded things to the state engine). Could you explain the use case for that specific example?

>  but many as you know also use NR instead of HA's capabilities and there attributes are not as pleasantly available so you either have to script it or create template sensors to use them instead.

That is an issue for Node-RED and cannot be a reason to add such features to Home Assistant. Please raise a feature request upstream instead. Nothing limits upstream from implementing any of this.

> So I personally use it for sun elevation, next sunset, sunrise, is it daylight according to my motion sensor and quite a bunch more including some custom components (which tend to put lots in attributes) which I both put in frontend but also use in automation.

Why not extend sun to provide those sensors instead?

> We should/could also extend this helper further in next step to allow for nested attributes such as in weather forecast as you might want to be able to show the rain for the next three days or something as such.

That sounds more like a trigger or condition that needs to be added for weather?

> And also consider probably to extend it into options of binary_sensor so one can choose between creating a sensor or a binary sensor depending on what you want to do with your new entity.

Right, that is actually one of my concerns with the proposal. `attribute_as_sensor`, could also provide binary sensors. Besides a naming thing, it also opens up other things. Next up would be a select, number, toggle, light helpers to extract an attribute with a service?


> Lots will still be attributes even if it's going to be even less than today eventually and this would be a simple UI friendly way to use them in a proper way.

So, we generally don't allow attributes in entities that provide states anymore. Most of those still present in our ecosystem, are more often legacies that can (and should) be replaced.

I think it makes more sense to focus on getting these attributes into their dedicated (binary)sensors, which everybody can benefit from directly, than adding a helper to get them out manually.

Additionally, this proposal makes it needed to handle all kinds of values, translations of these values and burdens user with device class, state classes and more of these things. Things, we should not burden the user with in the first place.

**Reply by gjohansson-ST** [reply]
> This example is odd, as the device class is a static property. Extracting those in a sensor with a helper should be disallowed and won't provide any value (besides raising memory and adding more unneeded things to the state engine). Could you explain the use case for that specific example?

My weather integration has an attribute for wind speed so by extracting it to it's own entity and setting the speed device class now I can switch between the different UoM's. Weather card provides this as option but an entity card would not (as it doesn't know what this attribute is).

> That sounds more like a trigger or condition that needs to be added for weather?

Maybe but I would think someone would also like to show this in frontend

> Right, that is actually one of my concerns with the proposal. `attribute_as_sensor`, could also provide binary sensors. Besides a naming thing, it also opens up other things. Next up would be a select, number, toggle, light helpers to extract an attribute with a service?

I was thinking quite a bit on this before making the PR in the first place as I renamed it from entity to sensor to at least by name limit it to such. Don't see a use case for the other platform and even less with a service and would not go there myself.

> So, we generally don't allow attributes in entities that provide states anymore. Most of those still present in our ecosystem, are more often legacies that can (and should) be replaced.

True in some cases but with some others they are build based on the attributes e.g. climate, weather as example.
Also here even if the logic is built around these attributes it can be beneficial to break them out into own sensors. Hence why some weather integrations even provide sensors in addition to the weather platform.

> I think it makes more sense to focus on getting these attributes into their dedicated (binary)sensors, which everybody can benefit from directly, than adding a helper to get them out manually.

While surely that is right it will take quite some time. This helper would/could help users directly.

> Additionally, this proposal makes it needed to handle all kinds of values, translations of these values and burdens user with device class, state classes and more of these things. Things, we should not burden the user with in the first place.

The requirements are name, entity and attribute. I would think it's pretty slim and simple unless someone wants to add more complexity to it.

I don't think this helper should change the base decision to move attributes into their own sensors within the integrations as far as possible. I would see it as a complement to the current helpers how to use and present your data.

**Reply by TomBrien** [reply]
I don't think we should make a helper to address bad practice in the code base or non-modernised integrations. This is a bandaid over the problem and may well cause a further bandaid to be needed down the line. We should instead fix the problems now by updating integrations

**Reply by frenck** [reply]
> My weather integration has an attribute for wind speed so by extracting it to it's own entity and setting the speed device class now I can switch between the different UoM's.

That integration can provide a wind sensor as well, out of the box, just working. We should invest time in such things instead of an helper...

> True in some cases but with some others they are build based on the attributes e.g. climate, weather as example.

They can provide additional sensors 🤷 

> While surely that is right it will take quite some time. This helper would/could help users directly.

I think it will slow down doing the right thing and, when wide used, will cause more breaking changes ahead (as these helpers will break when attributes removed/moved to a dedicated entity).

While this holds true for templates as well, those are considered more advanced use cases. Those users generally are better able to deal with such changes.

Additionally, if this is just for display in the frontend, time could be invested in making it possible to display attributes in the existing UI cards instead. No state has to be kept for those cases.

> I would see it as a complement to the current helpers how to use and present your data.

I'm not seeing that. What would really complement is adding those dedicated sensors instead.

**Reply by gjohansson-ST** [reply]
> I don't think we should make a helper to address bad practice in the code base or non-modernised integrations.

Who said that? There is loads of data in attributes that is going to stay as attributes I would think.
`sun` integration was perhaps a poor example but as it's the first one I had a problem with back in the days it's the one that came to mind discussing this.

**Reply by frenck** [reply]
> Who said that?

He said, so did I. The helper addresses/works around an issue of practice considered a bad practice nowadays.

**Reply by gjohansson-ST** [reply]
The intention was to make it simpler for users to show/use their own data. It's obvious you took the negative aspect that it can be used to "solve" bad coding which would be a necessary "setback" to allow this helper.

I still think the positive sides is beneficial enough that it should allow this helper but clearly I'm alone in thinking as such.

**Reply by TomBrien** [reply]
But should it not be us making the data easier to consume rather than requiring the user to do it themselves?

**Reply by gjohansson-ST** [reply]
That I would agree on hence my comment that this helper should not stop the work on getting the necessary attributes into their own sensors. As said it was not the purpose of this helper to bypass or stop that kind of work in progress.

**Reply by frenck** [reply]
>  As said it was not the purpose of this helper to bypass or stop that kind of work in progress.

Yet it will, it will reduce/slow down such progress. It will also make things break harder when entities actually do migrate. It above all isn't as user-friendly as it seems, as a user needs to think/care about things like device class & state classes all of a sudden.

For the reasons listed above by me, tom and jc2k, and based on the PR description and discussion above, I've left a negative recommendation on the implementation of this proposal.

Sorry 🙏 

../Frenck

**Reply by gjohansson-ST** [reply]
Well. Seems as all my smart ideas are hammered down lately but let's see about next one...

**Reply by frenck** [reply]
Sorry to hear that. Don't feel bad. That is where this repo/discussion can help.

**Reply by Kallb123** [reply]
How do we actually resolve this? It seems sensible to pull states out as sensors, but what happens with integrations that don't do it this way? There are developers/maintainers that disagree with adding sensors (which we don't need to delve into since it's already agreed above), hence users are the ones left with these attributes being more difficult to access.

Someone above said that it's up to us to make data easily accessible for the user. Without a way to enforce something like "no states in attributes", should we not give the user the tools to work around it?

**Comment by Jc2k** [comment, marked as answer]
I think we should break out attributes into their own sensors wherever it makes sense. If we made any architecture decision it would be to discourage the use of attributes that would be better as sensors/binary_sensor/etc.

**Reply by gjohansson-ST** [reply]
As written above. I don't question that and it's the path we should go.
The implementation of this helper would not change it.

**Reply by frenck** [reply]
It would change it, as perfectly summed up by @TomBrien here: <https://github.com/home-assistant/architecture/discussions/821#discussioncomment-3890169>

**Reply by hapklaar** [reply]
for anyone stumbling onto this and still needing attributes as entities, see [here](https://github.com/gjohansson-ST/attribute_as_sensor) My use case is for displaying on NSPanel.

**Comment by frenck** [comment]
We've discussed this path in various calls in the past too; in the end, we don't want to go this route. Our aim is to break out attributes instead.

Therefore, closing this architectural proposal.

../Frenck

**Reply by bliblabl8** [reply]
now, 4 months later, how is the problem solved? Is the user left alone with the issue for now and it will be just solved in years instead of a way shorter time?

**Reply by Junkbite** [reply]
will never be solved. thanks for the authoratative decision

**Comment by martydingo** [comment]
@frenck I would use something like this to split out the source attribute of a media player, as so I can use the lovelace conditional cards to display the correct media player entity based on the source of the media player.

I'd have to go about this by adding a text helper, and when the source changes, update the text helper. The ability to break attributes into entities on their own would create a cleaner pathway to this solution.

**Comment by fantnhu** [comment]
I monitor at least 10 exchange rates (opening price, current price, movement ...), I also monitor the phases of the Moon (phase, distance, visibility, tilt, next event ...), I also monitor the name days (primary, secondary, tomorrow ... - Hungary - ), the bills to be paid (amount, date of issue, payment deadline ...) and many more. So should I create a separate new template detector for each value, because there are no attributes in Helper? Instead of ~200 sensors, I should use 600-800. This is amazing.
